Scan extension apis

sachinthegreen opened the pull request:
Add scan extension APIs to provide controlled scan capability.

sachinthegreen committed:
net: wifi: Modify option and argument for scan type

Modify the way the scan type option is passed to the wifi scan command.
This makes it flexible to add more scan options.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Added scan extension support for frequency bands

Added scan extension to support scanning individual Wi-Fi bands or
combinations thereof.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Add scan extension for active channel dwell times

Add scan extension to control time spent on active scan channels.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Add scan extension for passive channel dwell times

Add scan extension to control scanning time spent on channels where
passive scanning is used.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Add scan extension for filtering SSIDs

Add scan extension to scan only for specific SSIDs.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -444,34 +445,57 @@ static int cmd_wifi_disconnect(const struct shell *sh, size_t argc,
 	return 0;
 }
 
+static int __wifi_scan_args_to_params(const struct shell *sh,
```
No need to have underscores in this private function

sachinthegreen replied:
Fixed.

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -96,6 +97,41 @@ static void scan_result_cb(struct net_if *iface, int status,
+		return -EINVAL;
+	}
+
+	band = strtok(scan_bands_str, ",");
```
Hmm, I am not sure about using strtok here, I recall its use should be avoided because its error prone. Do we need to separate elements with , in the first place?

sachinthegreen replied:
> Do we need to separate elements with , in the first place?

Considering the multitude of new options added (in subsequent commits), delimiter based parsing is working out to be better and more user friendly than positional parsing.

sachinthegreen committed:
net: wifi: Add scan extension for limiting scan results

Add scan extension to allow limiting the scan results to a user defined
maximum number. This parameter is intended to return results with the
best RSSI. It cannot be counted upon to limit the scan times since the
Wi-Fi chips might have to scan all the channels to find APs with the
best RSSI values across all channels.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Add scan extension for specifying channels

Add scan extension to allow limiting the scanning to specific channels.
Only 20 MHz channel specifications are allowed presently.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Add scan command option to print help

Add a scan command option to print the help text for the scan command.
This is to facilitate ease of using the command (since now there are
many different options added to the scan command).

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
net: wifi: Validate scan parameters

Add defensive check to validate scan command parameters.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

rlubos commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -0,0 +1,393 @@
+ * @brief Utility functions to be used by the Wi-Fi subsytem.
+ */
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_wifi_utils, LOG_LEVEL_INF);
```
We shouldn't set a fixed log level, the module should either specify its own log level in Kconfig, or reuse one of the existing (would this fall under CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL umbrella?).

sachinthegreen replied:
Thanks. Yeah it would make sense to have it under the CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL umbrella. Updated now.

rlubos commented on include/zephyr/net/wifi_utils.h:
```diff
@@ -0,0 +1,82 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Header guard is missing.

sachinthegreen replied:
Fixed.

rlubos reviewed (commented):
@sachinthegreen Please have a look at the CI failures.

It seems to that there is some dependency problem, `net_private.h` includes `wifi_mgmt.h` unconditionally, so we include the header even if WIFI_MGMT is disabled. And in that case, `CONFIG_WIFI_MGMT_SCAN_SSID_FILT_MAX` would not be defined, causing build errors.

jukkar commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -0,0 +1,395 @@
+	uint16_t i = 0;
+	bool start = false;
+	bool end = false;
+	uint16_t valid_5g_chans[] = {32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 96, 100, 104, 108,
```
If we are not changing this, then we could mark it const

sachinthegreen replied:
Made this a const.

jukkar commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -0,0 +1,395 @@
+	uint16_t valid_5g_chans[] = {32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 96, 100, 104, 108,
+		112, 116, 120, 124, 128, 132, 136, 140, 144, 149, 153, 157, 159, 161, 163, 165,
+		167, 169, 171, 173, 175, 177};
+	uint8_t idx = 0;
```
This seems unnecessary assignment as we assing idx value in all branches below.

jukkar commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -0,0 +1,395 @@
+
+static bool wifi_utils_validate_chan_5g(uint16_t chan)
+{
+	uint16_t valid_5g_chans[] = {32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 96, 100, 104, 108,
```
Perhaps const here?

sachinthegreen replied:
Made this a static const.

jukkar reviewed (commented):
As a general note, please only set variables init value if you expecting a default value there. There is no need to set the value in definition if you almost immediately set a value to it.

sachinthegreen commented:
> As a general note, please only set variables init value if you expecting a default value there. There is no need to set the value in definition if you almost immediately set a value to it.

Thanks. Removed initializations wherever possible.

sachinthegreen committed:
net: wifi: Enhance help text for NET_L2_WIFI_MGMT

Changed Add to Enable in help text to convey the intended use of the
Kconfig NET_L2_WIFI_MGMT.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen committed:
scripts: coccinelle: Remove strtok_r symbol from checklist

strtok_r is part of GNU99, but default Zephyr is built with C99, so,
the GNU headers are not pulled in, even if we add "--std=gnu99" it
won't override the C99 standard.

As a workaround for build we redefine the prototype, but this causes a
coding guideline warning, so, for now disable this warning
for strtok_r.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen commented on include/zephyr/net/wifi_utils.h:
```diff
@@ -0,0 +1,33 @@
+ */
+
+/** @file
+ * @brief Utility functions to be used by the Wi-Fi subsytem.
```
This is not expected to be a separate module, just an extension to wifi_mgmt. Added this to the `wifi_mgmt` group.

kartben replied:
Right, but I guess where I was coming from was more along the lines of making the documentation for wifi_mgmt less "flat" and crowded (see https://builds.zephyrproject.io/zephyr/pr/60686/docs/doxygen/html/group__wifi__mgmt.html), and some grouping wouldn't hurt (text helpers, parser utils, ...). Right now, this "flatness" may IMO unnecessarily give an impression of Wifi_mgmt being a "catch all" management module (which I guess it is tho :)) and it would be great to help users navigate it more easily.
You could consider using `@name { ... }` sections if you're not comfortable with a group showing up as a "Module".

krish2718 replied:
> Right now, this "flatness" may IMO unnecessarily give an impression of Wifi_mgmt being a "catch all" management module (which I guess it is tho :)) and it would be great to help users navigate it more easily.
> You could consider using `@name { ... }` sections if you're not comfortable with a group showing up as a "Module".

In this specific case `wifi_utils` is core part of `wifi_mgmt`, so, even though we add a new group or name, it will be shown along with `wifi_mgmt` docs, so, IMHO, no need for a separate group.

kartben replied:
@krish2718 While I am not questioning it's a "core part" of wifi_mgmt, wouldn't something along the lines of the following help bring more structure to the reader? (BTW, shouldn't there be a consistent way to get human-readable strings, vs. the current mix of helper functions and string arrays?)

<img width="835" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/c3f8bbb6-56e8-4554-9649-945a8f8e19f6">

krish2718 replied:
Sure, it's definitely good to have. But so far in Zephyr, we had let doxygen create standard `C` groups (functions, macros, data structures etc), haven't done any manual grouping, so, requires a wider overhaul than this PR.

kartben replied:
> Sure, it's definitely good to have. But so far in Zephyr, we had let doxygen create standard `C` groups (functions, macros, data structures etc), haven't done any manual grouping, so, requires a wider overhaul than this PR.

I would disagree, there are many areas where `@name`d sections are being defined and they really help bringing clarity. See e.g https://docs.zephyrproject.org/latest/doxygen/html/group__gpio__interface.html.
You are right that it would be great to have a larger Zephyr-wide overhaul (I am actually doing a lot of Doxygen cleanup these days), but I don't see at as a valid reason for not adding grouping wherever it makes sense. Plus, in this particular instance, I think that doing the "grouping exercise" is also good to help surfacing inconsistencies in the API w.r.t how to get human-readable string for a given enum.

krish2718 replied:
Thanks for the link. Using categories is definitely helpful but all I am saying is we can take this up in another doc improvement PR. Of course we can add `wifi_utils` as a separate `@name` in this PR itself.

sachinthegreen replied:
Thanks both. I have added a separate @name section for wifi_utils.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -495,6 +497,16 @@ static int wifi_scan_args_to_params(const struct shell *sh,
+		case 'a':
+			val = atoi(optarg);
+
+			if ((val <= 5) || (val >= 1000)) {
```
These should be inclusive, also, as per 802.11-20202 these are for OBSS scan, normal scan doesn't seem to have any limits.

sachinthegreen replied:
Yes, even though the specification specifies it for OBSS scan they seem to be sensible even for non-OBSS scans, hence using them here.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1253,7 +1260,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		    "[-a, --dwell_time_active <val_in_ms>] : Active scan dwell time (in ms) on a channel. Range 5 - 1000 ms.\n"
-		    "[-p, --dwell_time_passive <val_in_ms>] : Passive scan dwell time (in ms) on a channel. Range 10 - 1000 ms.",
+		    "[-p, --dwell_time_passive <val_in_ms>] : Passive scan dwell time (in ms) on a channel. Range 10 - 1000 ms.\n"
+		    "[-s, --ssid <Comma separate list of SSIDs>] : SSID list to scan for.",
```
`ssids` to be inline with `bands`.

sachinthegreen replied:
Thanks. Also changed `chan` to `chans`

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -198,6 +198,9 @@ struct wifi_scan_params {
 	uint16_t dwell_time_passive;
+	/** Array of SSID strings to scan.
+	 */
+	char ssids[CONFIG_WIFI_MGMT_SCAN_SSID_FILT_MAX][WIFI_SSID_MAX_LEN];
```
better to protect this with `ifdef` of `if` as by default `WIFI_MGMT_SCAN_SSID_FILT` is disabled.

sachinthegreen replied:
`WIFI_MGMT_SCAN_SSID_FILT` is not actually disabling the feature as such and just provides an option to statically specify a SSID filter. However I see your point and have reduced the default value of `CONFIG_WIFI_MGMT_SCAN_SSID_FILT_MAX` to cut down on consumed memory in the default case.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -63,3 +63,40 @@ int wifi_utils_parse_scan_bands(char *scan_bands_str, uint8_t *band_map)
+	ssid = strtok_r(scan_ssids_str, ",", &ctx);
+
+	while (ssid) {
+		if (strlen(ssid) > WIFI_SSID_MAX_LEN) {
```
`WIFI_SSID_MAX_LEN` this doesn't include the NULL de-limiter, so, need to add `+ 1`

sachinthegreen replied:
This check is still valid since strlen does not consider the NULL terminator.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -198,6 +198,9 @@ struct wifi_scan_params {
 	uint16_t dwell_time_passive;
+	/** Array of SSID strings to scan.
+	 */
+	char ssids[CONFIG_WIFI_MGMT_SCAN_SSID_FILT_MAX][WIFI_SSID_MAX_LEN];
```
`WIFI_SSID_MAX_LEN + 1` as you are storing NULL delimiter.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -63,3 +63,40 @@ int wifi_utils_parse_scan_bands(char *scan_bands_str, uint8_t *band_map)
+		}
+
+		if (i >= CONFIG_WIFI_MGMT_SCAN_SSID_FILT_MAX) {
+			NET_INFO("Exceeded maximum allowed (%d) SSIDs. Ignoring SSIDs %s onwards",
```
NET_WARN

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -104,17 +104,22 @@ static inline const char *wifi_mfp_txt(enum wifi_mfp_options mfp)
+ */
 enum wifi_frequency_bands {
-	/** 2.4GHz band. */
+	/** 2.4 GHz band. */
```
it would be good to have these doxygen changes in a separate commit.

sachinthegreen replied:
Ideally yes, but I felt that since these were related to the band extension changes I could just do it in that context.

krish2718 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -88,6 +88,13 @@ config WIFI_MGMT_SCAN_SSID_FILT
+	default 0
+	range 0 65535
+	help
+	  Maximum number of scan results to return. 0 represents unlimited number of BSSIDs.
```
BSSes

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1261,7 +1272,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		    "[-p, --dwell_time_passive <val_in_ms>] : Passive scan dwell time (in ms) on a channel. Range 10 - 1000 ms.\n"
-		    "[-s, --ssid <Comma separate list of SSIDs>] : SSID list to scan for.",
+		    "[-s, --ssid <Comma separate list of SSIDs>] : SSID list to scan for.\n"
+		    "[-m, --max_bss <val>] : Max BSSIDs to scan for. Range 1 - 65535.",
```
0 - 65535 (though 0 is default, we can still include it in the range)

sachinthegreen replied:
This is intentional since default is 0 (for unlimited results) and user doesnt need to specify this option if he wants unlimited results.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -13,26 +13,202 @@ LOG_MODULE_REGISTER(net_wifi_utils, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
 /* Ensure 'strtok_r' is available even with -std=c99. */
 char *strtok_r(char *str, const char *delim, char **saveptr);
 
+static const uint16_t valid_5g_chans[] = {32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 96, 100,
```
`valid_5g_chans_20mhz`

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -13,26 +13,202 @@ LOG_MODULE_REGISTER(net_wifi_utils, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
+
+static bool wifi_utils_validate_chan_2g(uint16_t chan)
+{
+	if ((chan >= 1) && (chan <= 14)) {
```
what about regulatory? Should we catch and raise error or as mentioned in the doc, should we let driver/chip handle this? IMHO catching earlier might be better esp. as we have access to `reg_domain`.

Or we can remove all validations here and let driver/chip handle.

sachinthegreen replied:
This PR relies on the driver/chip to handle regulatory validations. However due to the discrete and non-contiguous channel values in 5 GHz we need to impose some checks, especially to derive channel ranges.

krish2718 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -95,6 +95,24 @@ config WIFI_MGMT_SCAN_MAX_BSS_CNT
+	   before/after hyphens.
+	   An example channel specification specifying channels in the 2.4 GHz and 5 GHz bands is
+	   as below:
+	   2:1,5,7,9-11 5:36-48,100,163-167
```
Shell help and parsing code mentions `_`, but this isn't mentioned here.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1273,7 +1282,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		    "[-s, --ssid <Comma separate list of SSIDs>] : SSID list to scan for.\n"
-		    "[-m, --max_bss <val>] : Max BSSIDs to scan for. Range 1 - 65535.",
+		    "[-m, --max_bss <val>] : Max BSSIDs to scan for. Range 1 - 65535.\n"
+		    "[-c, --chan <Comma separated list of channel ranges>] : Channels to be scanned. The channels have to be specified in the form band1:chan1,chan2_band2:chan3,..etc. The chan has to be a list of comma separated values where each value is either a single channel or a channel range specified as chan_start-chan_end. Each band channel set has to be separated by a _. E.g. a valid channel specifica ...
```
isn't `space` a better delimiter than `_`?

sachinthegreen replied:
`space` is resulting in a separate argv, hence need a different delimiter.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -13,26 +13,202 @@ LOG_MODULE_REGISTER(net_wifi_utils, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
+			if (i == 1) {
+				i++;
+			} else if (i == 2) {
+				i += 2;
```
IIUC, this becomes channel 4 which isn't valid in 6GHz right?

sachinthegreen replied:
Yeah thanks. Fixed this.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -100,3 +276,121 @@ int wifi_utils_parse_scan_ssids(char *scan_ssids_str,
+	bool valid_chan = false;
+
+	while (scan_chan_str[i] != '\0') {
+		if (scan_chan_str[i] == ':') {
```
negating this condition and using `continue` will save you indentation.

krish2718 commented on subsys/net/l2/wifi/wifi_utils.c:
```diff
@@ -100,3 +276,121 @@ int wifi_utils_parse_scan_ssids(char *scan_ssids_str,
+
+	while (scan_chan_str[i] != '\0') {
+		if (scan_chan_str[i] == ':') {
+			if ((i - band_str_start_idx) > WIFI_UTILS_MAX_BAND_STR_LEN) {
```
better to add a check for `i - band_str_start_idx <= 0`

kartben commented on include/zephyr/net/wifi_utils.h:
```diff
@@ -0,0 +1,116 @@
+/**
+ * @name Wi-Fi utility functions.
+ *
+ * Utility functions for the Wi-Fi substem.
```
```suggestion
 * Utility functions for the Wi-Fi subsystem.
```

kartben commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -174,7 +174,11 @@ enum net_event_wifi_cmd {
 
-/** Wi-Fi scan parameters */
+/**
+ * @brief Wi-Fi scan parameters structure
```
properly terminate brief desc 
```suggestion
 * @brief Wi-Fi scan parameters structure.
```

kartben commented on include/zephyr/net/wifi_utils.h:
```diff
@@ -0,0 +1,116 @@
+
+/** @file
+ *
+ * @brief Utility functions to be used by the Wi-Fi subsytem.
+ */
```
```suggestion
 * @brief Utility functions to be used by the Wi-Fi subsystem.
 */
```

kartben commented on include/zephyr/net/wifi_utils.h:
```diff
@@ -0,0 +1,116 @@
+ * values as a comma separated string and convert it to a bitmap. The string can
+ * use the following characters to represent the bands:
+ *
+ * 2: 2.4 GHz
+ * 5: 5 GHz
+ * 6: 6 GHz
+ *
```
```suggestion
 * - 2: 2.4 GHz
 * - 5: 5 GHz
 * - 6: 6 GHz
 *
```

kartben commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -184,6 +188,40 @@ struct wifi_scan_params {
+	/** Channel information array indexed on Wi-Fi frequency bands and channels within that
+	 * band.
+	 * E.g. to scan channel 6 and 11 on the 2.4 GHz band, channel 36 on the 5 GHz band:
+	 *	chan[WIFI_FREQ_BAND_2_4_GHZ][0] = 6;
+	 *	chan[WIFI_FREQ_BAND_2_4_GHZ][1] = 11;
+	 *	chan[WIFI_FREQ_BAND_5_GHZ][0] = 36;
```
```suggestion
	 * @code{c}
	 *     chan[WIFI_FREQ_BAND_2_4_GHZ][0] = 6;
	 *     chan[WIFI_FREQ_BAND_2_4_GHZ][1] = 11;
	 *     chan[WIFI_FREQ_BAND_5_GHZ][0] = 36;
	 * @endcode
```

kartben reviewed (dismissed):
a few Doxygen fixes

rlubos reviewed (dismissed):
LGTM

sachinthegreen committed:
net: wifi: Enhance wifi_mgmt documentation

Enhance wifi_mgmt group documentation by :
 - Hiding some internal entities i.e. the human-readable strings
   defined in the various static const char * const arrays, by
   marking them as @cond INTERNAL_HIDDEN blocks.
 - Moving the Wi-Fi management API page under the networking group and
   adding a human readable name.

Signed-off-by: Sachin D Kulkarni <sachin.kulkarni@nordicsemi.no>

sachinthegreen commented on include/zephyr/net/wifi.h:
Changed `Wifi_mgmt` to `Wi-Fi Management` and moved it to https://docs.zephyrproject.org/latest/doxygen/html/group__networking.html
